### ISSUE TITLE
'bleeding' of style between paragraph and block

### ISSUE THREAD

**Issue by pm100** [tui-rs-revival/ratatui#144]
TL;DR; Is it correct that:
- the style of the contained paragraph should set the style of its surrounding block. 
- the style set in the block (via block.style) sets the style for the blank space in the para even if the para has its own style set.

If I have a paragraph with a block around it then there are 4 combinations of setting styles for the text and box

      para         block
      -------      ------- 
      default     default
      explicit     default  
      default     explicit
      explicit     explicit      

by default I mean `Style::default`
by explicit i mean say `Style::default().bg(Color::Red).fg(Color::Yellow)`

I used
- `Style::default().bg(Color::Red).fg(Color::Yellow)` for the block in my tests (setting block.border_style)
- `Style::default().bg(Color::Blue).fg(Color::Green)` for the paragraph

In a modified version of the paragraph sample app. (source below)

My expectation was a follows (for each combo listed above)

- white text on black everywhere
- green on blue text in a white on black box
- white on black text in a yellow on red box
- green on blue text in a yellow on red box

What I get is

![image](https://user-images.githubusercontent.com/489231/233509394-fe7ceb0c-6c05-4866-9012-5142caa9f400.png)

I also ran the same code with block.style instead of block.border_style. Nothing in the docs explains the difference between them. So I wanted to see what happened

![image](https://user-images.githubusercontent.com/489231/233511984-7fccbb3a-ce22-47dc-8b1d-50af8a1b6ef3.png)

This seems to be suggesting that block.style is intended to set the overall style of the para/block combo (para inherits from block), whereas block.border_style only sets the stye of the block. Fine.

But note that in the second case of both screen shots the para style has overriden the default of the block. This is a surprise

Also note the odd fact that although the para is explicity set to blue on green in the last case (of the last screen shot) the style set in the block style 'wins' in the 'blank' space

are these 2  intended behaviors?

I also tried with styled text rather that styling the paragraph (para always set to `Style::default`)

![image](https://user-images.githubusercontent.com/489231/233510678-fbace31c-e1d5-497c-8b6b-49732d60e4bd.png)

(There are only 2 combos now since the para style is always set to default but the text is set to white on blue)

Note the big difference for the second one. The style of the text is not propagated to the containing block.

and for completeness sake , style text with the block style set via block.border_style

![image](https://user-images.githubusercontent.com/489231/233511096-0e058691-3e98-49d5-8a97-883176d9fa2a.png)

here is the modified code of the paragraph sample

```fn ui<B: Backend>(f: &mut Frame<B>, app: &App) {
    let size = f.size();

    // Words made "loooong" to demonstrate line breaking.
    let s = "Veeeeeeeeeeeeeeeery    loooooooooooooooooong   striiiiiiiiiiiiiiiiiiiiiiiiiing.   ";
    let mut long_line = s.repeat(usize::from(size.width) / s.len() + 4);
    long_line.push('\n');

    //   let block = Block::default().style(Style::default()); //.bg(Color::White).fg(Color::Black));
    // f.render_widget(block, size);

    let chunks = Layout::default()
        .direction(Direction::Vertical)
        .margin(5)
        .constraints(
            [
                Constraint::Percentage(25),
                Constraint::Percentage(25),
                Constraint::Percentage(25),
                Constraint::Percentage(25),
            ]
            .as_ref(),
        )
        .split(size);

    let text = vec![Spans::from("This is a line ")];

    let create_block = |title, dflt| {
        Block::default()
            .borders(Borders::ALL)
            .title(Span::styled(
                title,
                Style::default().add_modifier(Modifier::BOLD),
            ))
            .border_style(if dflt {
                Style::default()
            } else {
                Style::default().bg(Color::Red).fg(Color::Yellow)
            })
    };

    let paragraph = Paragraph::new(text.clone())
        .style(Style::default())
        .block(create_block("dflt dflt", true))
        .alignment(Alignment::Left);
    f.render_widget(paragraph, chunks[0]);

    let paragraph = Paragraph::new(text.clone())
        .style(Style::default().bg(Color::Blue).fg(Color::Green))
        .block(create_block("para = Gr/Bl bl = dflt", true))
        .alignment(Alignment::Left);
    f.render_widget(paragraph, chunks[1]);
    let paragraph = Paragraph::new(text.clone())
        .style(Style::default())
        .block(create_block("para = dflt, block = Y/red", false))
        .alignment(Alignment::Left);
    f.render_widget(paragraph, chunks[2]);
    let paragraph = Paragraph::new(text.clone())
        .style(Style::default().bg(Color::Blue).fg(Color::Green))
        .block(create_block("Para = b/g block = y/red ", false))
        .alignment(Alignment::Left);
    f.render_widget(paragraph, chunks[3]);
}
```

## Environment

Add a description of the systems where you are observing the issue. For example:
- OS: Windows
- Terminal Emulator: windows terminal
- Font: ?
- Crate version: 0.20.1 (cloned current head)
- Backend: crossterm

Why do I care about ths. Because I am writing a widget that greys out a paragraphs text temporarily. If I do that when the surrounding block is set to `Style::default` then it gets greyed out too. And there is no way for me to inspect the block to look at its style (there is no get_style method)

**Comment by sophacles** [comment]
So I think the first thing to do is explain what the code currently does so we can decide if it's a bug.

To start, the Block is not just the borders around the paragraph, it's the whole box, and the paragraph is "in" the block. 

To begin with, the entire area of the block (including those where the paragraph will be drawn) are set to `Block.style`.  To do so it calls `Cell.set_style()` [here](https://github.com/tui-rs-revival/ratatui/blob/main/src/buffer.rs#L42-L52)  - it's important to note that if the the colors aren't set in the style (fg == None or bg == None  as set by `Style::default()`)  that any styling set by containing widgets (or other means) won't be removed. 

Then the boders are rendered - for this only the border cells are drawn, and the `Cell.set_style()` call is done on those, with `Block.border_style` as the style.

Rendering the block is the first thing that happens when you draw the paragraph. After all the above is finished, `Paragraph.render_widget()` calls `Cell.set_style()` when it renders the actual paragraph.

So with the styles you set, when you call `Style::default()` you get a style with `fg == None` and `bg == None`.  The behavior of `None` there is that whatever was there before remains, as you say it "bleeds through". This is the behavior of `Style` in a lot of places, not just `Cell.set_style` - it's intentionally additive. 

There is a way to "undo" the stylings set on a Cell. For colors this is `Color::Reset`.  For modifiers, Style maintains a list of "modifiers to add" and one of "modifiers to remove". These are all combined for you with the constructor `Style::reset()`, which will return the cell to whatever is the terminal default. (once you have constructed a style with `Style::reset()` you can then add background, foreground, modifers, etc as you see fit).

So in your code above, if you set the border style like this in your `create_block` closure:

```
            .border_style(if dflt {
                Style::reset()
            } else {
                Style::default().bg(Color::Red).fg(Color::Yellow)
            })
```
(note the `true` branch is `Style::reset()` now), I think it gets what you want.

**Comment by pm100** [comment]
My problem is explained in the footnote that you probably didnt get to (it was a long bug report)

I am adding a new feature to tui-textarea. My new feature needs to gray the text temporarily

 The API  interface for tui-textarea is that the caller supplies
- some text (optionally)
- some styling 
- an optional Block definition, if not given the widget uses a naked paragraph. This is not often done like that. Usually a Block::default is used

So I cannot control what the Block is set up as, my caller made that choice already. My test app was setup to try to exercise all the different combinations of block settings that might have been passed in because I got mighty confused as to what was happening

I cannot inspect the passed in Block because there is no option to read a Block's styles (style or border_style) , they are write only. If I could read them I could juggle their settings around to temporarily set the Block style

Still seems odd to me that settings the 'child' style (the para) should override the 'parent' (the block) style.

**Comment by sophacles** [comment]
Oh I see, I mixed up where you had control of the code, and were trying to control the inner region with the block. 

You can get part of the way there by doing the pargraph styling like this: `Paragraph.sytle(Style::reset().patch(passed_style))`, but that only adds correct behavior for `para = b/g bl = y/red`. 

Unfortunately the blue green/blue green is harder. I still don't think it's a bug. In my walk thru above since I was looking at what you can do with a block, i omitted that the paragraph sets its whole area  first. That makes sense to me because the paragraph is just using the block for decoration, the block is part of the paragraph, that's why the paragraph owns it. 

Since you're already making a widget, I'd reccomend leaning into it a bit. Instead of trying to do everything with the paragraph, just render the block yourself, and render a "naked" paragraph inside of it.. You can even do it inline as a widget pretty easily: https://gist.github.com/sophacles/7cf4bb8181485e5c418b0f0438881737

**Comment by pm100** [comment]
Thanks for the work on this. All I am trying to do is gray out the text in the paragraph for a small feature for the widget (which is not mine BTW). Changing the whole way it renders will make the PR look highly intrusive. I hoped that maybe massaging the style of the text directly rather than the paragraph would help. It does not

The main problem here is that there is no specification of what various things mean. So the codes behavior at present is thus correct since theres nothing to contradict its behavior :-(

There is no documentation about
- what Style::default() means. Does it mean, 'I really dont care make, it right depending on where its rendered'. Or does it mean - 'use the systems current setting for my terminal'
- how about if I dont specify a style, does it mean the same thing
- who is the winner in the relationship between block and para. Its seems that if either of the says default then the explicit style in the other wins
- is there any way to say 'use the system defined colors' . I think not. 
- what does Style::Reset mean, is it different from style.reset
- an easy doc fix is to explain the difference between block.style and block.border_style

By system defined I mean how my terminal looks when I am not running anything

Anyway I will look to see how intrusive the change will be doing it the way you suggest

TY

**Comment by sophacles** [comment]
I agree that the docs could be better. This is always true - I'm kind of being detailed here to supplement the docs until someone (me, you, random readers of this ticket) gets a chance to update the docs and submit a PR.  

That being said, there is documentation about what `Style::default()` means:

> Style let you control the main characteristics of the displayed elements.
> ...
> It represents an incremental change. If you apply the styles S1, S2, S3 to a cell of the terminal buffer, the style of this cell will be the result of the merge of S1, S2 and S3, not just S3.

It's an incremental change  - if you don't specify a change it stays the same.

(also:
> The default implementation returns a Style that does not modify anything. If you wish to reset all properties until that point use `Style::reset()`
)

from: https://docs.rs/ratatui/latest/ratatui/style/struct.Style.html

> is there any way to say 'use the system defined colors' . I think not.

Use `Style::reset()` to use 'system defined colors'. Please see quoted docs section and my first reply.

> who is the winner in the relationship between block and para. Its seems that if either of the says default then the explicit style in the other wins

Paragraph is given ownership of a Block by calling the `.block()` method. The paragraph draws in this order (for some more details please refer to my first reply):

1. set the style of the area it is given for drawing to: pararaph.style (or Style::default() if none is provided - this means don't change anything)
2. draw block - set the style of the entire area to the block. (style means merge, add the changes in) Then draw the borders if they specced, and apply that style (per the same merging rules as always).
3. draw text in the inner area of the block. Set styles from the Spans to the cells that text is drawn in. Styles, as always merge. 

Remember, when you merge in `Style::reset()` it returns to "system defined colors", and adds any further styling to a blank slate.

Paragraph render order could be documented better. (The whole module really)

> an easy doc fix is to explain the difference between block.style and block.border_style

Absolutely.

**Comment by pm100** [comment]
BTW - I meant what is Color::Reset not Style::Reset

**Comment by sophacles** [comment]
A quick check shows that the above ordering applies to most widgets the execptions are:

Block - because blocks don't have blocks.

Sparkline - skips step 1 (is this a bug?)

clear - it just resets every cell in the given area.

**Comment by pm100** [comment]
woohoo - your render the block and para separately works and its not that big of a change

Still think that graying out text in a block para should be a one liner :-)

I definitely think its a mistake that you cannot inspect a Block, its a read only object

**Comment by pm100** [comment]
FYI here it is now working . Thanks for your help

the feature is placeholder text (like html placeholder) a gray hint telling you what to type. The hint only appears when the text is empty

![tui](https://user-images.githubusercontent.com/489231/233757947-0ab89694-24f9-4899-a384-07a95074e64a.gif)

**Comment by mindoodoo** [comment]
I'll close this as it seems resolved, nice work @sophacles and nice feature @pm100 !
